File: 0791-custom-sort-string/0791-custom-sort-string.rs
use std::collections::HashMap;

impl Solution {
    pub fn custom_sort_string(order: String, s: String) -> String {
        let mut res = String::new();
        let mut mp: HashMap<char, i32> = HashMap::new();
        
        for ch in s.chars() {
            *mp.entry(ch).or_insert(0) += 1;
        }
        
        for ch in order.chars () {
            if let Some(cnt) = mp.get_mut(&ch) {
                while *cnt > 0 {
                    res.push(ch);
                    *cnt -= 1;
                }
             }
        }
        
        for (ch, cnt) in mp {
            for _ in 0..cnt {
                res.push(ch);
            }
        }
        res
    }
}

File: 0791-custom-sort-string/0791-custom-sort-string.cpp
class Solution {
public:
    string customSortString(string order, string s) {
        string res;
        unordered_map<char, int> mp;
        // get the counts of each char in s
        for (char ch : s) mp[ch]++;
        // check for the chars in 's' that are present in 'order'
        for (char ch : order) {
            if (mp.find(ch) != mp.end()) {
                while (mp[ch]-- > 0) {
                    res += ch;
                } 
            }
        }
        // for chars in 's' that are not in 'order'
        for (auto& [ch, cnt] : mp) {
            while (cnt-- > 0) {
                res += ch;
            }
        }
        return res;
    }
};